src/ray/raylet/scheduling/policy/bundle_scheduling_policy.cc

```diff
@@ -16,21 +16,62 @@
 
 namespace {
 
+/// Return true if scheduling this bundle (with resource_request) will exceed the
+/// max cpu fraction for placement groups.
+///
+/// \param node_resources The resource of the current node.
+/// \param resource_request The requested resources for the current bundle.
+/// \param max_cpu_fraction_per_node Highest CPU fraction the bundles can take up.
+/// \param available_cpus_before_scheduling Available CPUs on this node before
+///   scheduling the curernt bundles. It is used to calculate how many CPUs are
+///   allocated by the current bundles so far. It will help us figuring out
+///   the total CPU allocation from the current bundles for this node.
 bool AllocationWillExceedMaxCpuFraction(const ray::NodeResources &node_resources,
                                         const ray::ResourceRequest &resource_request,
-                                        double max_cpu_fraction_per_node) {
+                                        double max_cpu_fraction_per_node,
+                                        double available_cpus_before_scheduling) {
+  if (max_cpu_fraction_per_node == 1.0) {
+    // Allocation will never exceed the threshold if the fraction == 1.0.
+    return false;
+  }
+
   auto cpu_id = ray::ResourceID::CPU();
-  auto remaining_cpus = node_resources.available.Get(cpu_id).Double() -
-                        resource_request.Get(cpu_id).Double();
-  auto total_allocated_cpus = node_resources.total.Get(cpu_id).Double() - remaining_cpus;
+  auto total_cpus = node_resources.total.Get(cpu_id).Double();
+
+  // Calculate max_reservable_cpus
   auto max_reservable_cpus =
       max_cpu_fraction_per_node * node_resources.total.Get(cpu_id).Double();
 
   // If the max reservable cpu < 1, we allow at least 1 CPU.
   if (max_reservable_cpus < 1) {
     max_reservable_cpus = 1;
   }
-  return total_allocated_cpus > max_reservable_cpus;
+
+  // We guarantee at least 1 CPU is excluded from the placement group
+  // when max_cpu_fraction_per_node is specified.
+  if (max_reservable_cpus > total_cpus - 1) {
+    max_reservable_cpus = total_cpus - 1;
+  }
+
+  // Get the sum of all cpu allocations of past bundles.
+  FixedPoint cpus_for_existing_bundles(0);
```

Comment:
i guess the confusion comes from `existing_bundles`... can you name it to `cpus_allocated_for_placement_groups`?